libbeat/esleg/eslegclient/connection.go

```diff
@@ -407,6 +407,9 @@ func (conn *Connection) execHTTPRequest(req *http.Request) (int, []byte, error)
 		req.Host = host
 	}
 
+	// TODO: workaround for output reloading leaking FDs until context.WithCancel is used on transport dialer instead
+	req.Header.Set("Connection", "close")
```

Comment:
Before I made this workaround/hack, I tried two different approaches to fix the FD leak the "right way":

1. Just as there is an `http.Request` function (that we currently use in the `eslegclient` code), there is also a `http.RequestWithContext` function. I replaced our usage with the latter one, created a `context.WithCancel` inside `eslegclient.NewConnection` and used the cancel function returned to cancel the context from the `eslegclient.(*Connection).Close()` method. This is essentially the same fix that @urso tried in #10599. However, when I used `lsof` to monitor FDs being used by my Filebeat process, I noticed that this change had no effect!

2. So then, thinking about it some more, I realized the problem wasn't so much on canceling the HTTP requests but it was more to do with closing the underlying TCP connection. To fix this the "right way", I tried threading `context.Context` through the various `Dialer`s (which would allow us to call `DialContext` instead of `Dial` in the right places). This turned out to be quite a complicated and large refactoring! So instead, I settled for this workaround here. I think we do need to undertake this refactoring at some point but I'm not sure it's worth blocking progress at the moment?

Reply:
This would force the TCP connection to be dropped after each bulk request. Internally the go HTTP package uses a connection pool (actually the `Transport` type does) and keeps the HTTP session alive in order to send more requests. Would be nice if we could keep this.

This makes me wonder if this is really a leak, or just an open connection that eventually gets closed in a few minutes by the go runtime. The [`Transport`](https://pkg.go.dev/net/http?tab=doc#Transport) type has a few settings for controlling the connection pool. For example `IdleConnTimeout` (which we must configure). On shutdown one can also call `CloseIdleConnections`.
I think we should always configure a timeout (and expose it as a setting) + call `CloseIdleConnections` (best effort) approach to close connections early.

We also have some helpers for working with contexts in this package: https://pkg.go.dev/github.com/elastic/go-concert/ctxtool. This allows us to hook up contexts with custom functions, plain go channels, or merge cancelation from multiple contexts.

Reply:
Thanks @urso. 

> This makes me wonder if this is really a leak, or just an open connection that eventually gets closed in a few minutes by the go runtime.

Yes, good point. I will run my tests for longer and see if the connection FDs eventually get released.

> I think we should always configure a timeout (and expose it as a setting) + call CloseIdleConnections (best effort) approach to close connections early.

++ to this approach, will implement.

> We also have some helpers for working with contexts in this package: https://pkg.go.dev/github.com/elastic/go-concert/ctxtool. This allows us to hook up contexts with custom functions, plain go channels, or merge cancelation from multiple contexts.

I will look into this as well, but probably part of a follow up PR.

Reply:
++ Will all the above that @urso mentioned :)

Reply:
Implemented a new setting in `eslegclient.ConnectionSettings` named `IdleConnTimeout`. The name matches the setting on`http.Transport`. Default value is 1 minute.

Also implemented calling `CloseIdleConnections()` from `eslegclient.(*Connection).Close()`.

Net result is that we don't see FDs sticking around in the test.